engine/src/main/scala/cromwell/engine/workflow/tokens/TokenPool.scala

```diff
@@ -0,0 +1,49 @@
+package cromwell.engine.workflow.tokens
+
+import java.util.UUID
+
+import cromwell.core.JobExecutionToken
+import JobExecutionToken.JobExecutionTokenType
+import cromwell.engine.workflow.tokens.TokenPool.TokenPoolPop
+
+import scala.language.postfixOps
+
+trait TokenPool {
+  def currentLoans: Set[JobExecutionToken]
+  def pop(): TokenPoolPop
+  def push(jobExecutionToken: JobExecutionToken): TokenPool
+}
+
+object TokenPool {
+
+  case class TokenPoolPop(newTokenPool: TokenPool, poppedItem: Option[JobExecutionToken])
+
+  def apply(tokenType: JobExecutionTokenType): TokenPool = {
+    tokenType.maxPoolSize map { ps =>
+      val pool = (1 to ps toList) map { _ => JobExecutionToken(tokenType, UUID.randomUUID()) }
+      SizedTokenPool(pool.toSet, Set.empty)
+    } getOrElse {
+      InfiniteTokenPool(tokenType, Set.empty)
+    }
+  }
+}
+
+final case class SizedTokenPool(pool: Set[JobExecutionToken], override val currentLoans: Set[JobExecutionToken]) extends TokenPool {
+
+  override def pop(): TokenPoolPop = pool.toList match {
```

Comment:
The `toList`ing / `toSet` ing seems strange, perhaps just `pool.isEmpty` and getting the first element of the pool as an iterator?